### ISSUE TITLE
Possible race condition when accessing a Cache key

### ISSUE THREAD

**Issue by alessioarena** [dask/zict#69]
Unfortunately I'm currently unable to work on a minimum reproducible example, but I could follow up later with one.
However, if my diagnosis is correct it should not be hard to reproduce.

**Describe the issue**:

When attempting to access a Cache key in a dask worked (distributed) system I sometimes encounter the following error:
```python
File ~/./xarray_utilities/xrutils/indexing.py:135, in _sort_vals()
    133 # if we pass the indices values retrieve those
    134 if idx_vals is not None:
--> 135     idx_vals = dask.distributed.get_worker().data[idx_vals]
    136     idx = idx_vals[idx]
    137 return val, idx

File /env/lib/python3.8/site-packages/zict/buffer.py:104, in __getitem__()
    102 def __getitem__(self, key: KT) -> VT:
    103     if key in self.fast:
--> 104         return self.fast[key]
    105     elif key in self.slow:
    106         return self.slow_to_fast(key)

File /env/lib/python3.8/site-packages/zict/lru.py:76, in __getitem__()
     74 result = self.d[key]
     75 self.i += 1
---> 76 self.heap[key] = self.i
     77 return result

File /env/lib/python3.8/site-packages/heapdict.py:33, in __setitem__()
     30 @doc(dict.__setitem__)
     31 def __setitem__(self, key, value):
     32     if key in self.d:
---> 33         self.pop(key)
     34     wrapper = [value, key, len(self)]
     35     self.d[key] = wrapper

File /usr/lib/python3.8/_collections_abc.py:801, in pop()
    799     return default
    800 else:
--> 801     del self[key]
    802     return value

File /env/lib/python3.8/site-packages/heapdict.py:82, in __delitem__()
     80     parentpos = (wrapper[2] - 1) >> 1
     81     parent = self.heap[parentpos]
---> 82     self._swap(wrapper[2], parent[2])
     83 self.popitem()

File /env/lib/python3.8/site-packages/heapdict.py:71, in _swap()
     69 def _swap(self, i, j):
     70     h = self.heap
---> 71     h[i], h[j] = h[j], h[i]
     72     h[i][2] = i
     73     h[j][2] = j

IndexError: list index out of range
```

As you can see what I'm attempting to do is simply `worker.data[key]`, noting that `key` is the same for all threads in the worker.
This issue seems similar to https://github.com/dask/distributed/issues/2727 and using a Lock around the key access seems solving the problem.
Is this the expected behaviour?

**Anything else we need to know?**:

The main motivation to do this is to have a prescattered dataset, similarly to what `client.scatter` does.
However my implementation (that I will share soon for review with the dask community) it is resilient to workers being restarted or joining after the `scatter` call.

The key has been added using the dask API `worker.update_data` method

**Environment**:

- Dask version: 2022.9.1
- Python version: '3.8.10 (default, Jun 22 2022, 20:18:18) \n[GCC 9.4.0]'
- Operating System: Linux
- Install method (conda, pip, source): pip

**Comment by jrbourbeau** [comment]
Thanks for raising an issue @alessioarena. cc @crusaderky for visibility as he knows `distributed`s `SpillBuffer` well

**Comment by alessioarena** [comment]
> I'm very concerned about you trying to read from Worker.data directly - regardless of zict. If the key you're accessing is a direct dependency of your current task, you already have its value in the parameters. If it's not, you have no guarantee that it's going to be found on the worker to begin with. What makes you think otherwise?

Effectively I put it there using dask worker API. The current problem I'm facing is running an `xr.apply_ufunc` task in a dask_gateway context that requires pre-scattering data to the cluster.
Using the `client.scatter` method does not work as I cannot pass the Future to apply_ufunc in a way that is captured correctly by the scheduler. More on this, whatever way I used to pass the Future using `apply_ufunc`, this will be sent directly to the worker and as the Future includes the Client object and the worker does not know about `gateway` connections the task will fail. The error I receive is the same as [this](https://stackoverflow.com/questions/70775315/scattering-data-to-dask-cluster-workers-unknown-address-scheme-gateway).

As I cannot use a Future object to communicate this data to the scheduler/worker, the solution I came up with is to use a WorkerPlugin to send the data and register it in the `worker.data` cache with a known key, then retrieve it using that key.

At the moment I'm protecting that `worker.data[key]` call using the `worker._lock` threading Lock object. This "mostly" works, although I still occasionally see this error I reported and I'm not sure why

This is the full code at the moment
```Python
from distributed.diagnostics.plugin import WorkerPlugin
from distributed import Future, get_client, get_worker, Lock
import pickle, os
import logging
import threading
import uuid

class DataAccessor():
    """Small helper to access data broadcasted using the BroadcastData plugin
    
    Parameters
    ----------
    key : str
        key to identify the data broadcasted. This is typically the `name` property of BroadcastData
    """
    def __init__(self, key):
        self.key = key

    @property
    def data(self):

        worker = get_worker()
        data = None
        with worker._lock:
            if self.key in worker.data:
                data = worker.data[self.key]
        if data is None:
            raise KeyError("Requested key {0} was not found in worker {1}".format(self.key, worker.address))

        return data

    @property
    def future(self):
        return Future(self.key, client=get_client())


class BroadcastData(WorkerPlugin):
    """WorkerPlugin used to broadcast and load a python object as a distributed Future

    This is a more flexible alternative to client.scatter(data, broadcast=True) as data 
    transfer and Future generation are done reliable at worker setup, hence this makes sure that
    every worker has the data no matter if they are restarted or join the cluster after broadcasting
    the data

    To access the data please use the BroadcastData.data_accessor object and its .data property

    Parameters
    ----------
    data : any pickleable python object, bytes
        dataset to transfer
    client : Client
        Current Client you want the task to be register against

    Examples
    --------
    >>> broadcasted = BroadcastData(data, client)
    >>> client.register_worker_plugin(broadcasted)

    >>> broadcasted.future
    Future: broadcast_data status: finished, type: NoneType, key: broadcast_data-6a361cb0-f4ef-490d-a651-5eaab926920f
    >>> broadcasted.name
    broadcast_data-6a361cb0-f4ef-490d-a651-5eaab926920f
    >>> broadcasted.data_accessor
    < DataAccessor object at >
    """
   
    def __init__(self, data, client):
        if isinstance(data, bytes):
            self._data = data
        else:
            self._data = pickle.dumps(data)
        self.uuid = str(uuid.uuid4())
        self.name = "broadcast_data" + "-" + self.uuid
        self.client_id = client.id
        self.lock = Lock(self.name)
        self.data_accessor = DataAccessor(self.name)
        

    # def __del__(self):
    #     try:
    #         worker = get_worker()
    #         if self.name in worker.data:
    #             del worker.data[self.name]
    #     except ValueError:
    #         pass
        
    def delete(self):
        get_client().unregister_worker_plugin(self.name)
        future = self.future
        del future
        if self.lock.locked():
            self.lock.release()

    @property
    def future(self):
        return self.data_accessor.future
    @property
    def data(self):
        return self.data_accessor.data

    
    async def setup(self, worker):
        # init running environment
        logger = logging.getLogger("distributed.worker")
        self.name = "broadcast_data" + "-" + self.uuid

        # unpickle the data and load it in the worker's memory with a task key
        logger.info('loading data from plugin {0}'.format(self.name))
        data = {self.name: pickle.loads(self._data) }
        info = worker.update_data(data=data, report=False)
        
        # update the scheduler registry with this memory key
        # this is done only once to avoid the scheduler recognizing a duplicated key and trying to clean it up
        self.first_arrived = await self.lock.acquire(blocking=False)
        if self.first_arrived:
            who_has = {self.name: (worker.address,)}
            logger.info('Updating the scheduler registry with {0} for client ID {1}'.format(who_has, self.client_id))
            await worker.scheduler.update_data(
                who_has=who_has, 
                nbytes=info['nbytes'], 
                client=None
            )
            logger.info('Scheduler {0} updated succesfully'.format(worker.scheduler.address))

```

**Comment by alessioarena** [comment]
I may have figure out why I get still that error occasionally.. Coordinating data access among threads using a Lock does not prevent the parent worker from accessing data from the cache at the same time as a thread, for example if responding to a gather request.

Unfortunately I cannot see a way for me to resolve that data access at the worker level before passing the task to the thread, at least not using xr.apply_ufunc, so I may have to raise this issue there if there is no willingness to allow thread-safe data access at the zict level (I do understand that thread-safe was out of the original design of zict)

**Comment by crusaderky** [comment]
Manually injecting data into Worker.data is exceptionally bad practice. Things WILL break in every possible way - the way they just did for you is one of the many possibilities.

Please open an issue on dask/distributed, including a **minimal** reproducer, on why scatter() does not work for you.

**Comment by alessioarena** [comment]
I did open few issues and discussions around but I did not receive any response, for example:
https://github.com/pydata/xarray/issues/6803
https://github.com/pydata/xarray/discussions/6786

I think the limitation is in `dask.apply_gufunc` that is used internally by `xr.apply_ufunc`

Would be good for me to understand why using `worker.update_data` directly is exceptionally bad practice particularly relative to the code above. That obviously excluding the issue I already raised.

**Comment by crusaderky** [comment]
The worker state machine is a delicate interconnected set of many variables, which need to be coherent with the scheduler too. What you're trying to do is completely out of the dask design, and I think it would be a waste of developer time to support a use case that was never intended.

Closing as won't fix. In https://github.com/pydata/xarray/issues/6803 I showed the proper way to do it.

**Comment by alessioarena** [comment]
For completeness, here is a workaround to the problem I had.

I moved from having the data stored in the `Worker.data` Cache to accessing it from the plugin itself, stored in `Worker.plugins` dictionary.

## Advantages:
- does not incur in the problem described in this issue
- data is broadcasted and accessible to ALL workers no matter when they join the scheduler
- you can broadcast data when the cluster has 0 workers
- data is sent once to the scheduler, then replicated to every worker

## Disadvantages
- data access is slower than accessing it in the worker Cache
- you are required to access the data with a custom object (stored in the attribute `data_accessor`)

## Usage
```Python

# create and register the plugin
broadcasted = BroadcastData(data)
client.register_worker_plugin(broadcasted)

# retrieve the reference to the data
accessor = broadcasted.data_accessor

# example of how to access the data at runtime
def test(accessor):
    return accessor.data
client.submit(test, accessor)

```

## Code
```Python
class BroadcastDataAccessor():
    """Small helper to access data broadcasted using the BroadcastData plugin
    
    Parameters
    ----------
    key : str
        key to identify the data broadcasted. This is typically the `name` property of BroadcastData
    """
    def __init__(self, key):
        self.key = key
        self.logger = logging.getLogger("distributed.worker")

    @property
    def data(self):
        worker = get_worker()
        data = None
        self.logger.info('Acquiring data {0}'.format(self.key))
            
        if self.key in worker.plugins:
            data = worker.plugins[self.key].data

        if data is None:
            raise KeyError("Requested key {0} was not found in worker {1}".format(self.key, worker.address))
        self.logger.info(' Data {0} acquired'.format(self.key))
        return pickle.loads(data)




class BroadcastData(WorkerPlugin):
    """WorkerPlugin used to broadcast and load a python object as a distributed Future

    This is a more flexible alternative to client.scatter(data, broadcast=True) as data 
    transfer and Future generation are done reliable at worker setup, hence this makes sure that
    every worker has the data no matter if they are restarted or join the cluster after broadcasting
    the data

    To access the data please use the BroadcastData.data_accessor object and its .data property

    Parameters
    ----------
    data : any pickleable python object, bytes
        dataset to transfer
    client : Client
        Current Client you want the task to be register against

    Examples
    --------
    >>> broadcasted = BroadcastData(data)
    >>> client.register_worker_plugin(broadcasted)

    >>> broadcasted.name
    broadcast_data-6a361cb0-f4ef-490d-a651-5eaab926920f
    >>> broadcasted.data_accessor
    < DataAccessor object at >
    """
    def __init__(self, data):
        if isinstance(data, bytes):
            self._data = data
        else:
            self._data = pickle.dumps(data)
        self.uuid = str(uuid.uuid4())
        self.name = "broadcast_data" + "-" + self.uuid
        self.data_accessor = BroadcastDataAccessor(self.name)
        self.client_id = get_client().id
  
    def delete(self):
        get_client().unregister_worker_plugin(self.name)

    @property
    def data(self):
        return self._data

    async def setup(self, worker):
        # init running environment
        logger = logging.getLogger("distributed.worker")

        # data will be accessed using
        logger.info('data {0} succesfully broadcasted to worker {1}'.format(self.name, worker.name))

```
